### ISSUE TITLE
Dr. Racket becomes unresponsive when resizing the window

### ISSUE THREAD

**Issue by bmitc** [racket/racket#3560]
**What version of Racket are you using?**
7.9 [CS] on Windows 10

**What program did you run?**
Sometimes, but often, Dr. Racket will become unresponsive when resizing the window. When this occurs, sometimes it recovers if I leave it alone long enough and other times it remains unresponsive long enough that it must be killed and then re-opened. When it happens, the unresponsiveness happens immediately as I try to resize (i.e., when the border is clicked and a drag initiated) and not during a resize.

I have not been able to determine what occurs to make this happen, so unfortunately, the reproducibility instructions are to just use it long enough for it to happen. However, over the past day, I'd say it happened a dozen or so times. My primary use case was that a Microsoft Edge browser window was on the left with a PDF book open and a Dr. Racket window was on the right. Sometimes the two windows were in Windows 10's "multitasking" mode where there's the little bar in the middle that will resize both, but I would say the majority of time the two windows were simply overlaid. The unresponsiveness usually occurred during the latter situation, where I was resizing Dr. Racket horizontally on the left size. I don't know if the other window has anything to do with the bug, as I've had it occur otherwise as well.

**What should have happened?**
The Dr. Racket window should resize appropriately and be responsive.

**If you got an error message, please include it here.**
n/a

**Please include any other relevant details**
n/a

**Comment by bmitc** [comment]
I've actually had this happen a couple times today where Dr. Racket becomes unresponsive when clicking and dragging the vertical scrollbar. I am wondering now if it has more to do with a click and drag or just a drag than the resizing.

**Comment by mflatt** [comment]
Thanks for the report!

I've seen it with the scrollbar, too, but haven't yet reduced it to something small enough to figure out.

The problem is most likely related to GUI callbacks that have to be handled without Racket GC or thread switches. To handle those, the `racket/gui` makes an effort, but then suspends handling after a timeout, saving the current computation to resume later. So, the intent is that dragging the scrollbar will work, and if there's a timeout, then it will stop updating — but th GUI shouldn't freeze, and updating should recover as soon as you let go of the mouse. The same for window resizes.

Something is probably going wrong in the way the computation is supsended, or possible some queued callbacks pile up in a bad way.

**Comment by mflatt** [comment]
I think the problems I've seen may be addressed by a959c7f98857e2e651148f2acb43a7f7f6547cdf. I'd previously noticed the GC problem and categorized it as "minor" and "difficult to repair", but the correct categories were "major" and "easy to fix".

While the repair's effect is easiest to see on memory use, it could potentially affect other synchronization. So, if you get a chance to try a snapshot tomorrow (labeled 20201215 of later), that would be helpful.

**Comment by samth** [comment]
Note that there's already a snapshot with this fix at https://ci-snapshot.racket-lang.org/20201214-a959c7f988/index.html

**Comment by mflatt** [comment]
... but not for Windows

**Comment by bmitc** [comment]
Thanks, I'll try out a snapshot soon! I may not be able to get back to the project where I'm using Racket for a few days but will get to it by at least the weekend.

**Comment by sorawee** [comment]
@bmitc did you get a chance to try the snapshot version?

**Comment by bmitc** [comment]
I was able to download CS snapshot 7.9.0.13--2020-12-17 for Windows 10 over the weekend, but I haven't yet used it in a sustained enough manner (only for a few minutes). I'd like to be in the saddle, so to speak, for a couple of days of using it to have something to report back.

**Comment by bmitc** [comment]
I was finally able to spend some time back with Racket using snapshot `7.9.0.13--2020-12-17(52676db/a)`. I have not seen this issue reproduce with the snapshot. I did a lot of dragging around the window and some dragging of the vertical scrollbar (by clicking and dragging). I noticed some similar lagging as before, but DrRacket never locked up or became unresponsive. There was only a single lockup of DrRacket, but this was after running a file that had some test failures, and it was unresponsive in a different way. I think that's a separate unrelated issue, and it hasn't reproduced. Also, I am still able to reliably reproduce this in the released DrRacket 7.9 as a control.

So I think it's safe to consider this fixed. Thanks for the fix @mflatt!
